File: python/test_encryption.py
import sys
import getopt
import tea
import xtea
import crypto_help
import block_runner
def main(argv):
 # c=cipher, b=block mode, m=message, n=number of random chars to generate for message 
  block_cipher, block_mode, message = "", "", ""
  num_chars = 0
  verbose = False
  try: 
    opts, args = getopt.getopt(argv, "hc:b:m:n:v", ["help", "block_cipher=", "block_mode=", "message", "num_message_chars=", "verbose"])
  except:
    print "failed"
    usage()
    sys.exit(2)

  for opt, arg in opts:
    if opt in ("-h", "--help"):
      usage()
      sys.exit(2)
    elif opt in ("-c", "--block_cipher"):
      block_cipher = arg
      if block_cipher not in ("TEA", "XTEA"):
        print "Block cipher arg -c must be TEA or XTEA"
        sys.exit(2)
    elif opt in ("-b", "--block_mode"):
      block_mode = arg
      if block_mode not in ("ECB", "CBC"):
        print "Cipher mode must be ECB or CBC"
        sys.exit(2)
    elif opt in ("-m", "--message"):
      message = arg
      if "0" in message:
        print "Your message can't contain the character 0"
        sys.exit(2)
    elif opt in ("-n", "--num_message_chars"):
      num_chars = arg
    elif opt in ("-v", "--verbose"):
      verbose = True
  if message != "" and num_chars != 0:
    print "Can't enter both a message and a number of characters for a randomly generated message."
    sys.exit(2)
  if block_cipher == "":
    print "Please enter a block cipher -c TEA or -c XTEA"
    sys.exit(2)
  if block_mode == "":
    print "Please enter a cipher mode -b CBC or -b EBC"
    sys.exit(2)

  br = block_runner.BlockRunner()
  plaintext, key, iv, cipher, mode = "", "", "", "", ""
  key = crypto_help.gen_random_n_char_string(16)
  iv = crypto_help.gen_random_n_char_string(8)
  if message=="":
    #we're randomly generating one
    plaintext = crypto_help.gen_random_n_char_string(int(num_chars))
  else: 
    plaintext = message

  if block_cipher == "TEA":
    cipher = tea.TEA()
  elif block_cipher == "XTEA":
    cipher = xtea.XTEA()

  if block_mode == "ECB":
    encrypted = br.encrypt_ecb(cipher, key, plaintext)
    decrypted = br.decrypt_ecb(cipher, key, encrypted)
    if verbose:
      print "-- Encryption via %s using %s mode:" % (block_cipher, block_mode)
      print "Plaintext is %s" % plaintext
      print "Encrypted ciphertext is %s" % encrypted
      print "Ciphertext decrypts to %s" % decrypted
      print "Decrypted ciphertext and original plaintext are equivalent?   : %s" % (plaintext == decrypted)
  if block_mode == "CBC":
    encrypted = br.encrypt_cbc(cipher, key, plaintext, iv)
    decrypted = br.decrypt_cbc(cipher, key, encrypted, iv)
    if verbose:
      print "-- Encryption via %s using %s mode:" % (block_cipher, block_mode)
      print "Plaintext is %s" % plaintext
      print "Encrypted ciphertext is %s" % encrypted
      print "Ciphertext decrypts to %s" % decrypted
      print "Decrypted ciphertext and original plaintext are equivalent?   : %s" % (plaintext == decrypted)



  return

def usage():
  print "you need help"
if __name__=='__main__':
  main(sys.argv[1:])

File: rb/test_encryption.rb
#!/usr/bin ruby
$LOAD_PATH << '.'
require_relative 'cryptohelp'
require_relative 'tea'
require_relative 'xtea'
require_relative 'blockrunner'

# format: ruby test_encryption.rb CIPHER MODE [MESSAGE|NUM_CHARS] VERBOSE

cipher, mode, message, verbose, num_chars = "", "", false, false, false


cipher = ARGV[0]
mode = ARGV[1]
if (ARGV[2] =~ /\A\d+\z/ ? true : false) 
  num_chars = ARGV[2].to_i
else
  message = ARGV[2]
end
if ARGV[3]
  verbose = true
end




key = (0...16).map { (65 + rand(26)).chr}.join
iv = (0...8).map { (65 + rand(26)).chr}.join


if !message 
  original_plaintext = (0...num_chars).map { (65 + rand(26)).chr }.join
else 
  original_plaintext = message
end

if cipher=="TEA"
  if mode == "ECB"
    #TEA in ECB
    encrypted = BlockRunner.encrypt_ecb(TEA, original_plaintext, key)
    decrypted = BlockRunner.decrypt_ecb(TEA, encrypted, key).gsub('0', '')
    if verbose
      puts "\n--- Testing TEA in ECB Mode\n\n"
      puts "Plaintext is: \n#{original_plaintext}"
      puts "Ciphertext is:"
      puts "#{Crypto.convert_fixnum_to_string(encrypted)}"
      puts "Decrypted ciphertext: \n#{decrypted}"
    end

  else
    #tea in CBC
    encrypted = BlockRunner.encrypt_cbc(TEA, original_plaintext, key, iv)
    decrypted = BlockRunner.decrypt_cbc(TEA, encrypted, key, iv).gsub('0', '')
    if verbose
      puts "\n--- Testing TEA in CBC Mode\n\n"
      puts "Plaintext is: \n#{original_plaintext}"
      puts "Ciphertext is:"
      puts "#{Crypto.convert_fixnum_to_string(encrypted)}"
      puts "Decrypted ciphertext: \n#{decrypted}"
      puts original_plaintext.class
      puts decrypted.class
    end
  end
else
    #cipher is XTEA
    if mode == "ECB"
      #TEA in ECB
      encrypted = BlockRunner.encrypt_ecb(XTEA, original_plaintext, key)
      decrypted = BlockRunner.decrypt_ecb(XTEA, encrypted, key).gsub('0', '')
      if verbose
        puts "\n--- Testing XTEA in ECB Mode\n\n"
        puts "Plaintext is: \n#{original_plaintext}"
        puts "Ciphertext is:"
        puts "#{Crypto.convert_fixnum_to_string(encrypted)}"
        puts "Decrypted ciphertext: \n#{decrypted}"
      end
    else
    #tea in CBC
    encrypted = BlockRunner.encrypt_cbc(XTEA, original_plaintext, key, iv)
    decrypted = BlockRunner.decrypt_cbc(XTEA, encrypted, key, iv).gsub('0', '')
    if verbose
      puts "\n--- Testing XTEA in CBC Mode\n\n"
      puts "Plaintext is: \n#{original_plaintext}"
      puts "Ciphertext is:"
      puts "#{Crypto.convert_fixnum_to_string(encrypted)}"
      puts "Decrypted ciphertext: \n#{decrypted}"
      puts original_plaintext.class
      puts decrypted.class
    end
  end

end



